### ISSUE TITLE
inference code

### ISSUE THREAD

**Issue by albertchristianto** [jfzhang95/pytorch-deeplab-xception#85]
Hi, @jfzhang95 

it's a very good repository. I am a new at deep learning research. And how to only use this repository to infer 1 image ? if possible, can anyone share the inference code?

thank you very much.

best regards,
Albert

**Comment by leolle** [comment]
@jfzhang95 please take a look at the request.

**Comment by theevann** [comment]
Here is a very simple code for inference. Note that you need to have downloaded the checkpoint ""deeplab-resnet.pth" before and put it in the same directory (url availabe in git readme).

```python
from PIL import Image
import numpy as np

import torch
import torchvision.transforms as tr

from modeling.deeplab import DeepLab
from dataloaders.utils import decode_segmap

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

checkpoint = torch.load("deeplab-resnet.pth")

model = DeepLab(num_classes=21,
                backbone='resnet',
                output_stride=16,
                sync_bn=True,
                freeze_bn=False)

model.load_state_dict(checkpoint['state_dict'])
model.eval()
model.to(device)


def transform(image):
    return tr.Compose([
        tr.Resize(513),
        tr.CenterCrop(513),
        tr.ToTensor(),
        tr.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))
        ])(image)

torch.set_grad_enabled(False)

image = Image.open('image2.jpg')
inputs = transform(image).to(device)
output = model(inputs.unsqueeze(0)).squeeze().cpu().numpy()
pred = np.argmax(output, axis=0)

# Then visualize it:
decode_segmap(pred, dataset="pascal", plot=True)
```

**Comment by andeyeluguo** [comment]
I just test the code ,found the result is not so satisfying.

**Comment by albertchristianto** [comment]
Thank you @theevann for the response. Actually i have finished write the inference code by my self .
Below is the sample code that i used.
############################################################################
import argparse
import os
import numpy as np
import torch
import cv2

from modeling.sync_batchnorm.replicate import patch_replication_callback
from modeling.deeplab import *

from dataloaders.utils import decode_segmap

numClass={'pascal':21,
		  'coco':21,
		  'cityscapes':19}
						
parser = argparse.ArgumentParser(description="PyTorch DeeplabV3Plus Training")
parser.add_argument('--backbone', type=str, default='resnet',
					choices=['resnet', 'xception', 'drn', 'mobilenet'],
					help='backbone name (default: resnet)')
parser.add_argument('--out-stride', type=int, default=16,
					help='network output stride (default: 8)')
parser.add_argument('--dataset', type=str, default='cityscapes',
					choices=['pascal', 'coco', 'cityscapes'],
					help='dataset name (default: pascal)')
parser.add_argument('--sync-bn', type=bool, default=None,
					help='whether to use sync bn (default: auto)')
parser.add_argument('--freeze-bn', type=bool, default=False,
					help='whether to freeze bn parameters (default: False)')
parser.add_argument('--weightPath', type=str, default=None,
                        help='put the path to resuming file if needed')
parser.add_argument('--imgPath', type=str, default=None,
                        help='put the path to resuming file if needed')
parser.add_argument('--outPath', type=str, default=None,
                        help='put the path to resuming file if needed')
parser.add_argument('--no-cuda', action='store_true', default=
                        False, help='disables CUDA training')
args = parser.parse_args()
args.cuda = not args.no_cuda and torch.cuda.is_available()

nclass = numClass[args.dataset]

model = DeepLab(num_classes=nclass,
                        backbone=args.backbone,
                        output_stride=args.out_stride,
                        sync_bn=args.sync_bn,
                        freeze_bn=args.freeze_bn)
if args.cuda:
	model = torch.nn.DataParallel(model, device_ids=[0])
        patch_replication_callback(model)
        model = model.cuda()

weight_dict=torch.load(args.weightPath)
model.module.load_state_dict(weight_dict['state_dict'])
model.eval()

image = cv2.imread(args.imgPath)
image = image.astype(np.float32) / 255.
means = [0.485, 0.456, 0.406]
stds = [0.229, 0.224, 0.225]

image = image[:, :, ::-1]
for i in range(3):
	image[:, :, i] = image[:, :, i] - means[i]
	image[:, :, i] = image[:, :, i] / stds[i]
oriDim = image.shape
image = cv2.resize(image, dsize=(513,513))

image = torch.from_numpy(image.transpose((2, 0, 1)).astype(np.float32)).float().unsqueeze(0)

if args.cuda:
	image = image.cuda()
with torch.no_grad():
	output = model(image)
output = output.data.cpu().numpy()
prediction = np.argmax(output, axis=1)
segmap = decode_segmap(prediction[0], dataset=args.dataset)
segmap = (segmap*255).astype(np.uint8)
segmap = cv2.resize(segmap,dsize=(oriDim[1],oriDim[0]))
segmap = segmap[:, :, ::-1]
cv2.imwrite(args.outPath,segmap)
##############################################################################

**Comment by linkAmy** [comment]
@theevann Thanks for your inference code, it is very useful. But when I run your code, here comes the error: 

> size mismatch for decoder.last_conv.8.weight: copying a param of torch.Size([19, 256, 1, 1]) from checkpoint, where the shape is torch Size([21, 256, 1, 1]) in current model.
Note that I use the pretrained model "deeplab-resnet.pth", and keep all the params just as you are.

**Comment by theevann** [comment]
@linkAmy This is a problem with the number of class. You are using a checkpoint with 19 classes, so you should change the `num_classes` parameter in the call to instantiate the model :
```python
model = DeepLab(num_classes=19, # Change this line
                backbone='resnet',
                output_stride=16,
                sync_bn=True,
                freeze_bn=False)
```

**Comment by leedoge** [comment]
> Here is a very simple code for inference. Note that you need to have downloaded the checkpoint "deeplab-resnet.pth" before and put it in the same directory (url availabe in git readme).
> from PIL import Image
> import numpy as np
> 
> import torch
> import torchvision.transforms as tr
> 
> from modeling.deeplab import DeepLab
> from dataloaders.utils import decode_segmap
> 
> device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
> 
> checkpoint = torch.load("deeplab-resnet.pth")
> 
> model = DeepLab(num_classes=21,
>                 backbone='resnet',
>                 output_stride=16,
>                 sync_bn=True,
>                 freeze_bn=False)
> 
> model.load_state_dict(checkpoint['state_dict'])
> model.eval()
> model.to(device)
> 
> def transform(image):
>     return tr.Compose([
>         tr.Resize(513),
>         tr.CenterCrop(513),
>         tr.ToTensor(),
>         tr.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))
>     ])(image)
> 
> torch.set_grad_enabled(False)
> 
> image = Image.open('image.jpg')
> inputs = transform(image).to(device)
> output = model(inputs.unsqueeze(0)).squeeze().cpu().numpy()
> pred = np.argmax(output, axis=0)
> 
> # Then visualize it:
> decode_segmap(pred, dataset="pascal", plot=True)
> Note that this is a code for a model trained on the Pascal Voc dataset. If you want to use it with another dataset, you need to change the num_classes parameter and the normalization.
> For example, for Cityscapes, use num_classes=19, and change the normalization to:
>  tr.Normalize(mean=(0.2838, 0.3251, 0.2869))
> For visualization, you then need: decode_segmap(pred, dataset="cityscapes", plot=True)



> Here is a very simple code for inference. Note that you need to have downloaded the checkpoint "deeplab-resnet.pth" before and put it in the same directory (url availabe in git readme).
> from PIL import Image
> import numpy as np
> 
> import torch
> import torchvision.transforms as tr
> 
> from modeling.deeplab import DeepLab
> from dataloaders.utils import decode_segmap
> 
> device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
> 
> checkpoint = torch.load("deeplab-resnet.pth")
> 
> model = DeepLab(num_classes=21,
>                 backbone='resnet',
>                 output_stride=16,
>                 sync_bn=True,
>                 freeze_bn=False)
> 
> model.load_state_dict(checkpoint['state_dict'])
> model.eval()
> model.to(device)
> 
> def transform(image):
>     return tr.Compose([
>         tr.Resize(513),
>         tr.CenterCrop(513),
>         tr.ToTensor(),
>         tr.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))
>     ])(image)
> 
> torch.set_grad_enabled(False)
> 
> image = Image.open('image.jpg')
> inputs = transform(image).to(device)
> output = model(inputs.unsqueeze(0)).squeeze().cpu().numpy()
> pred = np.argmax(output, axis=0)
> 
> # Then visualize it:
> decode_segmap(pred, dataset="pascal", plot=True)
> Note that this is a code for a model trained on the Pascal Voc dataset. If you want to use it with another dataset, you need to change the num_classes parameter and the normalization.
> For example, for Cityscapes, use num_classes=19, and change the normalization to:
>  tr.Normalize(mean=(0.2838, 0.3251, 0.2869))
> For visualization, you then need: decode_segmap(pred, dataset="cityscapes", plot=True)

hello! I have run your code and find it very useful,thank you! but i find the output image is cropped,how can i get a result with the same size of my input!thank yooo very much!!!

**Comment by theevann** [comment]
Just remove these two lines:
```
tr.Resize(513),
tr.CenterCrop(513),
```

**Comment by beizhengren** [comment]
@theevann Hi, theevann thank you for providing the useful code.
I want to finetune on my own dataset with only 3 classes. How shold I do it?
Thanks

**Comment by theevann** [comment]
You should replace the last convolution layer of the decoder by a convolution layer with 3 channels.
Something like :
`self.model.decoder.last_conv[-1] = nn.Conv2d(256, 3, kernel_size=1, stride=1)`

**Comment by beizhengren** [comment]
@theevann Thanks a lot!

**Comment by rudtn0852** [comment]
@theevann  it works great before but i have a error now after change something ( i have no idea ) 

Exception has occurred: UnpicklingError
A load persistent id instruction was encountered,
but no persistent_load function was specified.
  File "/home/seg/pytorch-deeplab-xception/infer.py", line 12, in <module>
    checkpoint = torch.load("/home/seg/pytorch-deeplab-xception/checkpoint.pth.tar")

even i use "deeplab-resnet.pth" , same error occured. Do you have any idea ?
